### ISSUE TITLE
Win: Visual Studio 2019 Unresolved external symbols

### ISSUE THREAD

**Issue by kf185042** [acristoffers/CEF3SimpleSample#23]
Steps to duplicate:
1. Clone repo
2. Download the Windows txz file, extract to ...\CEF\Win
3. Start Visual Studio 2019 v16.9.1
4. Open the file "CEF3SimpleSample.sln"
5. VisualStudio prompts to "Open Configuration Manager" to correct setting.  Correct it and close (removes x64 tags).
6. Select Retarget projects (from v120 to v142), and keep old Windows SDK
7. Add "Additional Directory" to find **libcef_dll_wrapper.lib** in (.\CEF\Win)
8. Build output:

Build started...
```
1>------ Build started: Project: CEF3SimpleSample, Configuration: Release Win32 ------
1>ClientApp.obj : error LNK2001: unresolved external symbol _cef_string_utf16_clear
1>ClientApp.obj : error LNK2001: unresolved external symbol _cef_string_utf8_to_utf16
1>ClientApp.obj : error LNK2001: unresolved external symbol "bool __cdecl CefRegisterExtension(class CefStringBase<struct CefStringTraitsUTF16> const &,class CefStringBase<struct CefStringTraitsUTF16> const &,class scoped_refptr<class CefV8Handler>)" (?CefRegisterExtension@@YA_NABV?$CefStringBase@UCefStringTraitsUTF16@@@@0V?$scoped_refptr@VCefV8Handler@@@@@Z)
1>ClientHandler.obj : error LNK2001: unresolved external symbol "public: __thiscall cef::logging::LogMessage::~LogMessage(void)" (??1LogMessage@logging@cef@@QAE@XZ)
1>ClientHandler.obj : error LNK2001: unresolved external symbol "public: __thiscall cef::logging::LogMessage::LogMessage(char const *,int,int)" (??0LogMessage@logging@cef@@QAE@PBDHH@Z)
1>ClientHandler.obj : error LNK2001: unresolved external symbol _cef_get_min_log_level
1>ClientV8ExtensionHandler.obj : error LNK2001: unresolved external symbol _cef_string_utf16_cmp
1>ClientV8ExtensionHandler.obj : error LNK2001: unresolved external symbol _cef_string_utf16_to_utf8
1>ClientV8ExtensionHandler.obj : error LNK2001: unresolved external symbol "public: static class scoped_refptr<class CefV8Context> __cdecl CefV8Context::GetCurrentContext(void)" (?GetCurrentContext@CefV8Context@@SA?AV?$scoped_refptr@VCefV8Context@@@@XZ)
1>ClientV8ExtensionHandler.obj : error LNK2001: unresolved external symbol _cef_string_utf8_clear
1>main.obj : error LNK2001: unresolved external symbol "public: static bool __cdecl CefBrowserHost::CreateBrowser(class CefWindowInfo const &,class scoped_refptr<class CefClient>,class CefStringBase<struct CefStringTraitsUTF16> const &,class CefStructBase<struct CefBrowserSettingsTraits> const &,class scoped_refptr<class CefDictionaryValue>,class scoped_refptr<class CefRequestContext>)" (?CreateBrowser@CefBrowserHost@@SA_NABVCefWindowInfo@@V?$scoped_refptr@VCefClient@@@@ABV?$CefStringBase@UCefStringTraitsUTF16@@@@ABV?$CefStructBase@UCefBrowserSettingsTraits@@@@V?$scoped_refptr@VCefDictionaryValue@@@@V?$scoped_refptr@VCefRequestContext@@@@@Z)
1>main.obj : error LNK2001: unresolved external symbol _cef_string_utf16_set
1>main.obj : error LNK2001: unresolved external symbol "bool __cdecl CefInitialize(class CefMainArgs const &,class CefStructBase<struct CefSettingsTraits> const &,class scoped_refptr<class CefApp>,void *)" (?CefInitialize@@YA_NABVCefMainArgs@@ABV?$CefStructBase@UCefSettingsTraits@@@@V?$scoped_refptr@VCefApp@@@@PAX@Z)
1>main.obj : error LNK2001: unresolved external symbol "public: static class scoped_refptr<class CefCommandLine> __cdecl CefCommandLine::GetGlobalCommandLine(void)" (?GetGlobalCommandLine@CefCommandLine@@SA?AV?$scoped_refptr@VCefCommandLine@@@@XZ)
1>main.obj : error LNK2001: unresolved external symbol "int __cdecl CefExecuteProcess(class CefMainArgs const &,class scoped_refptr<class CefApp>,void *)" (?CefExecuteProcess@@YAHABVCefMainArgs@@V?$scoped_refptr@VCefApp@@@@PAX@Z)
1>main.obj : error LNK2001: unresolved external symbol "void __cdecl CefRunMessageLoop(void)" (?CefRunMessageLoop@@YAXXZ)
1>main.obj : error LNK2001: unresolved external symbol "void __cdecl CefShutdown(void)" (?CefShutdown@@YAXXZ)
1>C:\NcrDev\code\GitHub\CEF3SimpleSample\Release\CEF3SimpleSample.exe : fatal error LNK1120: 17 unresolved externals
1>Done building project "CEF3SimpleSample.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

**Comment by acristoffers** [comment]
The libraries probably need recompiling in more modern environments, but I'm not going to do it anytime soon. Sorry.
Go to cefbuild, download the newest source and compile it yourself, then replace it in the folder. This might be enough.

**Comment by acristoffers** [comment]
I still have no idea what is the problem, but I bumped the CEF versions in the project and the build works with CMake.

**Comment by JipBoesenkool** [comment]
Also having trouble with the latest version. For me the application did build but I could not run it. (opens and closes)
Trying the older version (v79.1.4) worked for me. On first sight building the latest version was missing a lot of files for me.

**Comment by acristoffers** [comment]
It seems some compilation flags changed, but I could not pinpoint what changed. On Windows, the mangled symbol names depend on the compilation flags used, so I bet that's what's happening.

About missing files, there is another issue complaining about this, and I realized that I had a wrong .gitgnore entry. It should be fixed now, I updated all include folders and now they have the new header files that were added to CEF since my last update.

**Comment by srhtgryaktuna** [comment]
> 1>------ Build started: Project: CEF3SimpleSample, Configuration: Release Win32 ------
> 1>ClientApp.obj : error LNK2001: unresolved external symbol _cef_string_utf16_clear
> 1>ClientApp.obj : error LNK2001: unresolved external symbol _cef_string_utf8_to_utf16
> 1>ClientApp.obj : error LNK2001: unresolved external symbol "bool __cdecl CefRegisterExtension(class CefStringBase<struct CefStringTraitsUTF16> const &,class CefStringBase<struct CefStringTraitsUTF16> const &,class scoped_refptr<class CefV8Handler>)" (?CefRegisterExtension@@YA_NABV?$CefStringBase@UCefStringTraitsUTF16@@@@0V?$scoped_refptr@VCefV8Handler@@@@@Z)
> 1>ClientHandler.obj : error LNK2001: unresolved external symbol "public: __thiscall cef::logging::LogMessage::~LogMessage(void)" (??1LogMessage@logging@cef@@QAE@XZ)
> 1>ClientHandler.obj : error LNK2001: unresolved external symbol "public: __thiscall cef::logging::LogMessage::LogMessage(char const *,int,int)" (??0LogMessage@logging@cef@@QAE@PBDHH@Z)
> 1>ClientHandler.obj : error LNK2001: unresolved external symbol _cef_get_min_log_level
> 1>ClientV8ExtensionHandler.obj : error LNK2001: unresolved external symbol _cef_string_utf16_cmp
> 1>ClientV8ExtensionHandler.obj : error LNK2001: unresolved external symbol _cef_string_utf16_to_utf8
> 1>ClientV8ExtensionHandler.obj : error LNK2001: unresolved external symbol "public: static class scoped_refptr<class CefV8Context> __cdecl CefV8Context::GetCurrentContext(void)" (?GetCurrentContext@CefV8Context@@SA?AV?$scoped_refptr@VCefV8Context@@@@XZ)
> 1>ClientV8ExtensionHandler.obj : error LNK2001: unresolved external symbol _cef_string_utf8_clear
> 1>main.obj : error LNK2001: unresolved external symbol "public: static bool __cdecl CefBrowserHost::CreateBrowser(class CefWindowInfo const &,class scoped_refptr<class CefClient>,class CefStringBase<struct CefStringTraitsUTF16> const &,class CefStructBase<struct CefBrowserSettingsTraits> const &,class scoped_refptr<class CefDictionaryValue>,class scoped_refptr<class CefRequestContext>)" (?CreateBrowser@CefBrowserHost@@SA_NABVCefWindowInfo@@V?$scoped_refptr@VCefClient@@@@ABV?$CefStringBase@UCefStringTraitsUTF16@@@@ABV?$CefStructBase@UCefBrowserSettingsTraits@@@@V?$scoped_refptr@VCefDictionaryValue@@@@V?$scoped_refptr@VCefRequestContext@@@@@Z)
> 1>main.obj : error LNK2001: unresolved external symbol _cef_string_utf16_set
> 1>main.obj : error LNK2001: unresolved external symbol "bool __cdecl CefInitialize(class CefMainArgs const &,class CefStructBase<struct CefSettingsTraits> const &,class scoped_refptr<class CefApp>,void *)" (?CefInitialize@@YA_NABVCefMainArgs@@ABV?$CefStructBase@UCefSettingsTraits@@@@V?$scoped_refptr@VCefApp@@@@PAX@Z)
> 1>main.obj : error LNK2001: unresolved external symbol "public: static class scoped_refptr<class CefCommandLine> __cdecl CefCommandLine::GetGlobalCommandLine(void)" (?GetGlobalCommandLine@CefCommandLine@@SA?AV?$scoped_refptr@VCefCommandLine@@@@XZ)
> 1>main.obj : error LNK2001: unresolved external symbol "int __cdecl CefExecuteProcess(class CefMainArgs const &,class scoped_refptr<class CefApp>,void *)" (?CefExecuteProcess@@YAHABVCefMainArgs@@V?$scoped_refptr@VCefApp@@@@PAX@Z)
> 1>main.obj : error LNK2001: unresolved external symbol "void __cdecl CefRunMessageLoop(void)" (?CefRunMessageLoop@@YAXXZ)
> 1>main.obj : error LNK2001: unresolved external symbol "void __cdecl CefShutdown(void)" (?CefShutdown@@YAXXZ)
> 1>C:\NcrDev\code\GitHub\CEF3SimpleSample\Release\CEF3SimpleSample.exe : fatal error LNK1120: 17 unresolved externals
> 1>Done building project "CEF3SimpleSample.vcxproj" -- FAILED.
> ========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
> ```


I got same error, has anyone to solved it?

**Comment by acristoffers** [comment]
No, no one submitted pull requests fixing it. The CMake build works, meaning this is some misconfiguration on the Visual Studio project, but I did not manage to pinpoint what changed.

**Comment by srhtgryaktuna** [comment]
Can you give me some information about how to build with CMake? Because i used CMake GUI but generates Visual Studio Solution files.

**Comment by srhtgryaktuna** [comment]
I sent an e-mail your gmail account, can you review ?

**Comment by acristoffers** [comment]
Yes, I saw it but didn't even have time to respond. I'm not looking forward to connecting to a stranger's computer. This example is meant to be a beginner's reference code to show how to call JS from C++ and C++ from JS, nothing else. I don't understand why so many people want to use it as scaffolding for their projects. I don't know what is wrong with the VS project because I not even use Windows, I just have a VM that I start when needed, which is almost never. Mostly to update this project. But I'm not going to waste a ton of time debugging Windows problems, I just don't have enough time for that. If someone finds out what is wrong and submits a pull request I'll gladly review and merge, but that's about it, I'm not going to put any more effort on fixing this.

**Comment by g-h-c** [comment]
See #27, which provides the build steps for VS 2019 on Windows

**Comment by acristoffers** [comment]
Fixed in 27782717a88fe3887e1cb5058c5961cf01b91fea.
